### PR TITLE
N-API bindings added

### PR DESCRIPTION
**Pull request by chm-diederichs**

### CODE REVIEW

File: macros.h [hunk 1]
```diff
@@ -1,44 +1,198 @@
...
+  int32_t name##_int32; \
+  if (napi_get_value_int32(env, val, &name##_int32) != napi_ok) { \
+    napi_throw_error(env, "EINVAL", "Expected number"); \
+    return NULL; \
+  } \
+  SN_THROWS(name##_int32 > 255, "expect uint8") \
+  unsigned char name = 0xff & name##_int32;
+
+#define SN_UINT32(name, val) \
+  int32_t name; \
```
Comment (mafintosh) [inline macros.h]: should be uint32?

File: macros.h [hunk 2]
```diff
@@ -1,44 +1,198 @@
...
+  if (napi_get_value_int32(env, val, &name) != napi_ok) { \
```
Comment (mafintosh) [inline macros.h]: get_value_uint32?

File: macros.h [hunk 3]
```diff
@@ -1,44 +1,198 @@
...
+  napi_typedarray_type name##_type; \
+  size_t name##_length; \
+  assert(napi_get_typedarray_info(env, (var), &name##_type, &name##_length, &name##_data, NULL, NULL) == napi_ok); \
+  uint8_t name##_width = typedarray_width(name##_type); \
+  SN_THROWS(name##_width == 0, "Unexpected TypedArray type") \
+  name##_size = name##_length * name##_width;
+
+#define SN_UINT8(name, val) \
+  int32_t name##_int32; \
+  if (napi_get_value_int32(env, val, &name##_int32) != napi_ok) { \
```
Comment (mafintosh) [inline macros.h]: uint?

File: binding.c [hunk 4]
```diff
@@ -11,24 +99,2109 @@ napi_value sn_randombytes_random (napi_env env, napi_callback_info info) {
...
+
+  return result;
+}
+
+napi_value sn_sodium_is_zero(napi_env env, napi_callback_info info) {
+  SN_ARGV(1, sodium_is_zero);
+
+  SN_ARGV_TYPEDARRAY(a, 0)
+
+  sodium_is_zero(a_data, a_size);
```
Comment (mafintosh) [inline binding.c]: this line is not needed

File: binding.c [hunk 5]
```diff
@@ -11,24 +99,2109 @@ napi_value sn_randombytes_random (napi_env env, napi_callback_info info) {
...
+
+napi_value sn_sodium_pad(napi_env env, napi_callback_info info) {
+  SN_ARGV(3, sodium_pad);
+
+  SN_ARGV_TYPEDARRAY(buf, 0)
+  SN_ARGV_UINT32(unpadded_buflen, 1)
+  SN_ARGV_UINT32(blocksize, 2)
+
+  SN_THROWS(unpadded_buflen > (int) buf_size, "unpadded length cannot exceed buffer length")
+  SN_THROWS(blocksize > (int) buf_size, "block size cannot exceed buffer length")
```
Comment (mafintosh) [inline binding.c]: remove casts after fixing macro above

File: binding.c [hunk 6]
```diff
@@ -11,24 +99,2109 @@ napi_value sn_randombytes_random (napi_env env, napi_callback_info info) {
...
+
+napi_value sn_sodium_unpad(napi_env env, napi_callback_info info) {
+  SN_ARGV(3, sodium_unpad);
+
+  SN_ARGV_TYPEDARRAY(buf, 0)
+  SN_ARGV_UINT32(padded_buflen, 1)
+  SN_ARGV_UINT32(blocksize, 2)
+
+  SN_THROWS(padded_buflen > (int) buf_size, "unpadded length cannot exceed buffer length")
+  SN_THROWS(blocksize > (int) buf_size, "block size cannot exceed buffer length")
```
Comment (mafintosh) [inline binding.c]: remove cast after fixing macro

Comment (chm-diederichs) [reply binding.c]: fixed by e4065a8

File: binding.c (hunk 5 above)
Comment (chm-diederichs) [reply binding.c]: fixed by e4065a8

File: macros.h (hunk 3 above)
Comment (chm-diederichs) [reply macros.h]: fixed by e4065a8

File: macros.h (hunk 2 above)
Comment (chm-diederichs) [reply macros.h]: fixed by e4065a8

File: binding.c (hunk 4 above)
Comment (chm-diederichs) [reply binding.c]: fixed by 129571f

File: binding.c [hunk 7]
```diff
@@ -11,24 +99,2109 @@ napi_value sn_randombytes_random (napi_env env, napi_callback_info info) {
...
+  SN_RETURN_BOOLEAN(crypto_pwhash_scryptsalsa208sha256_str_verify(str_data, pwd_data, pwd_size))
+}
+
+// CHECK: returns 1, 0, -1
+napi_value sn_crypto_pwhash_scryptsalsa208sha256_str_needs_rehash (napi_env env, napi_callback_info info) {
+  SN_ARGV(3, crypto_pwhash_scryptsalsa208sha256_str_needs_rehash)
+
+  SN_ARGV_TYPEDARRAY(hash, 0)
+  SN_ARGV_UINT64(opslimit, 1)
+  SN_ARGV_UINT64(memlimit, 2)
```
Comment (mafintosh) [inline binding.c:927]: Since we cannot pass in 64bit numbers, I think we need to pass in uint32s here (cc @emilbayes )

File: binding.c [hunk 8]
```diff
@@ -11,24 +99,2109 @@ napi_value sn_randombytes_random (napi_env env, napi_callback_info info) {
...
+  SN_ARGV_TYPEDARRAY(client_pk, 2)
+  SN_ARGV_TYPEDARRAY(client_sk, 3)
+  SN_ARGV_TYPEDARRAY(server_pk, 4)
+
+  SN_THROWS(client_pk_size != crypto_kx_PUBLICKEYBYTES, "client public key must be 32 bytes")
+  SN_THROWS(client_sk_size != crypto_kx_SECRETKEYBYTES, "client secret key must be 32 bytes")
+  SN_THROWS(server_pk_size != crypto_kx_PUBLICKEYBYTES, "server public key must be 32 bytes")
+
+  SN_THROWS(tx_size != crypto_kx_SESSIONKEYBYTES && tx_data != NULL, "trasnmitting key buffer must be 32 bytes or null")
+  SN_THROWS(rx_size != crypto_kx_SESSIONKEYBYTES && tx_data != NULL, "receiving key buffer must be 32 bytes or null")
```
Comment (mafintosh) [inline binding.c]: rx_data right?

File: binding.c [hunk 9]
```diff
@@ -11,24 +99,2109 @@ napi_value sn_randombytes_random (napi_env env, napi_callback_info info) {
...
+  crypto_kdf_keygen(key_data);
+
+  return NULL;
+}
+
+napi_value sn_crypto_kdf_derive_from_key (napi_env env, napi_callback_info info) {
+  SN_ARGV(4, crypto_kdf_derive_from_key)
+
+  SN_ARGV_TYPEDARRAY(subkey, 0)
+  SN_ARGV_UINT64(subkey_id, 1)
```
Comment (mafintosh) [inline binding.c:1266]: should prob also be a 32bit uint, cc @emilbayes

File: binding.c [hunk 10]
```diff
@@ -11,24 +99,2109 @@ napi_value sn_randombytes_random (napi_env env, napi_callback_info info) {
...
+  napi_value callback;
+  SN_STATUS_THROWS_VOID(napi_get_reference_value(env, req->cb, &callback), "");
+
+  napi_value return_val;
+  SN_STATUS_THROWS_VOID(napi_call_function(env, global, callback, 1, argv, &return_val), "");
+  SN_STATUS_THROWS_VOID(napi_delete_reference(env, req->cb), "");
+  SN_STATUS_THROWS_VOID(napi_delete_reference(env, req->output_ref), "");
+  SN_STATUS_THROWS_VOID(napi_delete_reference(env, req->pwd_ref), "");
+  SN_STATUS_THROWS_VOID(napi_delete_reference(env, req->salt_ref), "");
+  SN_STATUS_THROWS_VOID(napi_delete_async_work(env, req->task), "");
```
Comment (mafintosh) [inline binding.c]: none of these can throw afaik as it's async. should just be asserts

File: binding.c [hunk 11]
```diff
@@ -11,24 +99,2109 @@ napi_value sn_randombytes_random (napi_env env, napi_callback_info info) {
...
+}
+
+napi_value sn_crypto_pwhash_async (napi_env env, napi_callback_info info) {
+  SN_ARGV(7, crypto_pwhash_async)
+
+  SN_ARGV_BUFFER_CAST(unsigned char *, output, 0)
+  SN_ARGV_BUFFER_CAST(char *, pwd, 1)
+  SN_ARGV_BUFFER_CAST(unsigned char *, salt, 2)
+  SN_ARGV_UINT64(opslimit, 3)
+  SN_ARGV_UINT64(memlimit, 4)
```
Comment (mafintosh) [inline binding.c:1634]: uint32? cc @emilbayes

File: binding.c [hunk 12]
```diff
@@ -11,24 +99,2109 @@ napi_value sn_randombytes_random (napi_env env, napi_callback_info info) {
...
+
+  SN_STATUS_THROWS(napi_queue_async_work(env, req->task), "")
+
+  return NULL;
+}
+
+typedef struct async_pwhash_str_verify_request {
+  napi_ref hash_ref;
+  char * hash_data;
+  napi_re
```
Comment (mafintosh) [inline binding.c]: should be static

**Comment by mafintosh** [comment]
@chm-diederichs can you add crypto_stream_xor_init, update, final also. I don't know if that one is in sodium, but was in the nan version of we use that one a lot

File: binding.c [hunk 13]
```diff
@@ -11,24 +99,2109 @@ napi_value sn_randombytes_random (napi_env env, napi_callback_info info) {
...
+  return result;
+}
+
+napi_value sn_crypto_sign_keypair(napi_env env, napi_callback_info info) {
+  SN_ARGV(2, crypto_sign_keypair)
+
+  SN_ARGV_TYPEDARRAY(pk, 0)
+  SN_ARGV_TYPEDARRAY(sk, 1)
+
+  SN_THROWS(pk_size != crypto_sign_PUBLICKEYBYTES, "public key must be 32 bytes")
```
Comment (mafintosh) [inline binding.c]: can you change the error messages to be "public key must be crypto_sign_PUBLICKEYBYTES bytes"

Comment (mafintosh) [reply binding.c]: for all the errors, so people look at the constants

Comment (emilbayes) [reply binding.c]: I think it should be even more precise like: "pk must be crypto_sign_PUBLICKEYBYTES bytes", ie. the "ARG must be CONST bytes"

File: binding.c [hunk 14]
```diff
@@ -11,24 +99,2109 @@ napi_value sn_randombytes_random (napi_env env, napi_callback_info info) {
...
+
+napi_value sn_crypto_sign_open(napi_env env, napi_callback_info info) {
+  SN_ARGV(3, crypto_sign_open)
+
+  SN_ARGV_TYPEDARRAY(message, 0)
+  SN_ARGV_TYPEDARRAY(signed_message, 1)
+  SN_ARGV_TYPEDARRAY(pk, 2)
+
+  SN_THROWS(message_size != signed_message_size - crypto_sign_BYTES, "message buffer must be 64 bytes shorter than input")
+  SN_THROWS(signed_message_size < crypto_sign_BYTES, "signed message must be at least 64 bytes")
```
Comment (mafintosh) [inline binding.c]: should this not be != also?

File: binding.c (hunk 13 above)
Comment (mafintosh) [reply binding.c]: 👍

File: binding.c [hunk 15]
```diff
@@ -11,24 +99,2109 @@ napi_value sn_randombytes_random (napi_env env, napi_callback_info info) {
...
+  return NULL;
+}
+
+typedef struct async_pwhash_str_verify_request {
+  napi_ref hash_ref;
+  char * hash_data;
+  napi_ref pwd_ref;
+  const char * pwd_data;
+  size_t pwd_size;
+
```
Comment (mafintosh) [inline binding.c:2359]: for all the _init functions we need to expose the sizeof(struct) they need so you know in JS how much to alloc.

can you add those also?

```
SN_EXPORT_UINT32(crypto_hash_sha256_sizeof, sizeof(the-struct))
```

Comment (mafintosh) [reply binding.c:2359]: sorry, saw them below

File: binding.c (hunk 14 above)
Comment (chm-diederichs) [reply binding.c]: `sm` can be any length > `crypto_sign_BYTES` here since `m` is prepended

File: binding.c (hunk 12 above)
Comment (chm-diederichs) [reply binding.c]: get a `cannot combine with previous 'static' declaration specifier` error here

File: binding.c (hunk 8 above)
Comment (chm-diederichs) [reply binding.c]: fixed by 05082e9

File: binding.c (hunk 13 above)
Comment (chm-diederichs) [reply binding.c]: fixed by 05082e9

File: binding.c (hunk 10 above)
Comment (chm-diederichs) [reply binding.c]: fixed by f228156

File: binding.c (hunk 11 above)
Comment (chm-diederichs) [reply binding.c:1634]: fixed by 3e2ae75

File: binding.c (hunk 9 above)
Comment (chm-diederichs) [reply binding.c:1266]: fixed by 3e2ae75

File: binding.c (hunk 7 above)
Comment (chm-diederichs) [reply binding.c:927]: fixed by 3e2ae75

Comment (chm-diederichs) [reply binding.c:927]: decided to revert back to uint64 in c708bbe

File: binding.c [hunk 16]
```diff
@@ -1,7 +1,110 @@
 #include <node_api.h>
+#include <string.h>
 #include <assert.h>
-#include "macros.h"
 #include <sodium.h>
+#include "macros.h"
+
+uint8_t typedarray_width(napi_typedarray_type type) {
```
Comment (mafintosh) [inline binding.c]: shoudl be static

File: macros.h [hunk 17]
```diff
@@ -1,44 +1,249 @@
...
+
+#define SN_ARGV_BUFFER_CAST(type, name, index) \
+  napi_value name##_argv = argv[index]; \
+  SN_BUFFER_CAST(type, name, name##_argv)
+
+#define SN_ARGV_BUFFER_TO_BYTE_TAG(type, name, index) \
+  napi_value name##_argv = argv[index]; \
+  napi_typedarray_type name##_type; \
+  size_t name##_length; \
+  uint8_t * name##_data; \
```
Comment (mafintosh) [inline macros.h]: nit: *name##_data instead (no whitespace)

File: binding.c (hunk 16 above)
Comment (chm-diederichs) [reply binding.c]: fixed by d6114c6

File: macros.h (hunk 17 above)
Comment (chm-diederichs) [reply macros.h]: fixed by d6114c6

**Pull request merged**